### DISCUSSION TITLE
Another SQL Join Question

### DISCUSSION THREAD

**Discussion by fadethegap** [supabase/supabase#5799] [Questions] [answered]
I have looked through the posts on joins and their linked articles but haven't found exactly what I'm looking for.

I was hopeful that I could write a rpc and drop the below sql in, but there is no recordset return value, so I only get one record when I call it.

```
-- Get the count of downloads for a specific contributor in a date range
SELECT s.title, COUNT(s.title)
FROM(
  SELECT  DISTINCT a.title, p.email
  FROM asset_downloads ad
  LEFT OUTER JOIN cms_asset a ON ad.asset_id = a.id
  LEFT OUTER JOIN profile p ON ad.user_id = p.id
  WHERE a.contributor_id = '1a2b3c4d-e6eb-4a3c-b2f9-abc123abc'
  AND ad.calc_date BETWEEN '2022/03/01' AND '2022/03/31') s 
GROUP BY s.title
```
I know I can arrive at the same result through javascript and various calls to the dB, but I get it done in one shot with this.

Does anyone have a recommendation?

-- The purpose of the above query: Contributors submit assets into our repository, and users download them. Contributors get paid based on the number of unique downloads via a given user,  and users often download an asset more than once. The above subquery obtains the number of unique downloads, and the outer query counts them.

**Comment by GaryAustin1** [comment, marked as answer]
For a setof record return you need to use the sql editor window.  The function UI does not handle return arrays.

Something like this (function was for someone else):
```
drop function fetch_users();

create or replace function fetch_users(uid text)
language sql returns setof record as $$
select * from user_data t1 where t1."userId"
not in (select match_request_to from matches_data where
match_request_from in(uid));
$$;
```
https://github.com/supabase/supabase/discussions/5668   Helping someone else with same issue.

**Reply by fadethegap** [reply]
I modified the above to create a simple function:

```
create or replace function fetch_assets()
language sql returns setof record as $$
select * from cms_asset;
$$;
```
Which yielded:
```
syntax error at or near "setof"
```
I'm super pumped that this feature is a possibility. I just need to crack open what is going on. Does anything pop out to you?

**Reply by fadethegap** [reply]
Aha! This worked:

```
create or replace function fetch_assets(aid)
returns setof record as $$
select * from cms_asset;
$$ language sql; 
```

As always, thank you @GaryAustin1!

**Reply by GaryAustin1** [reply]
Sorry, I changed the example I gave... return has to come before the language.
